### ISSUE TITLE
How to slow down typing speed?

### ISSUE THREAD

**Issue by otaconn23** [ideasman42/nerd-dictation#67]
I would like to reduce the typing speed and limit text prediction period.  It should look like 75-90 words per minute.  A major issue is that I will dictate a sentence then it will backspace a big portion of that sentence replacing wit with something else that isn't as accurate the first time around. Please advise what file I should change. Thank you.

model: vosk-model-en-us-0.22

**Comment by ideasman42** [comment]
Firstly the delay between key-strokes could be made into a command line option.

As for the issue with the first option being better than what is re-typed, I think this is out of our control. 

As far as I can see: VOSK uses a statistical so the assumption is generally that hearing more words will produce a more accurate dictation of what your saying (generally I find this is true). If with more information the results are worse, I suppose there may be some internal buffer which is used for predicting word order, which could be reduced so dictation is closer to the initially selected words. I'd need to check on the VOSK API to see if it's possible.

Failing that, it's possible to "reset" the recognizer which could be used to accept a partial match
```
diff --git a/nerd-dictation b/nerd-dictation
index 27a9bf2..53065ae 100755
--- a/nerd-dictation
+++ b/nerd-dictation
@@ -846,6 +846,7 @@ def text_from_vosk_pipe(
 
     # Use code to delay exiting, allowing reading the recording buffer to catch-up.
     code = 0
+    retype_ok = False
 
     if idle_time > 0.0:
         idle_time_prev = time.time()
@@ -875,13 +876,23 @@ def text_from_vosk_pipe(
 
             if ok:
                 json_text = rec.Result()
+            elif not retype_ok:
+                json_text = rec.PartialResult()
+                json_data = json.loads(json_text)
+                text = json_data["partial"]
+                if text:
+                    json_text = rec.FinalResult()
+                    rec.Reset()
+                    ok = True
+
+            if ok:
                 json_text_partial_prev = ""
                 json_data = json.loads(json_text)
                 text = json_data["text"]
                 assert isinstance(text, str)
                 if text:
                     handle_fn_wrapper(text, False)
-            else:
+            elif retype_ok:
                 # Only for comparison, to detect change.
                 # if use_timeout:
                 json_text = rec.PartialResult()
```

This complicates the main loop a bit though, so I'd rather avoid this hack if necessary.

----

Note that accepting the first guess might be handy for issuing commands where re-issuing the command doesn't make sense.

**Comment by otaconn23** [comment]
Perhaps there is a way to let VOSK do its analyzing but block nerd-dictation.py (xdotool) backspacing and re-typing. This would prevent commands from being re-issued without messing with VOSK. Furthermore, to indicate to the user that processing is complete, nerd-dictation.py (xdotool) can key "Space" x2 (also programmed as a separator).

**Comment by ideasman42** [comment]
I don't think it can be this simple, as the partial result may not contain words later re-typed in the final result as there is a delay between each evaluation that accepts new audio.

This is why I think it's necessary to force reset the dictation model once a partial evaluation is found, although that's not a perfect solution either as it's possible a partial evaluation is given even when you have not finished speaking, although in practice I found it works fairly well.

But ideally there would be a way to always accept the first guess and not change result.

Another option could be to never accept partial results, only final ones, then at least there will never be any back-spacing. Although if for some reason the results are worse for you, that wont help.

**Comment by otaconn23** [comment]
You're a gem, thank you. I'm a little new to git though I'm familiar with Linux. Can you perhaps guide me on how to implement your code suggestion to evaluate the "reset" method please? Thank you!

**Comment by ideasman42** [comment]
Manually applying patches isn't so convenient, so committed the changes to a branch: `test-no-retyping` - if you checkout this branch in git, it should always use the partial result, then move on to evaluating whatever you say afterwards.

**Comment by flesler** [comment]
> Another option could be to never accept partial results, only final ones

@ideasman42 this here. Can we have a --flag to do this? Only keep the final one?
